nipoppy_new/nipoppy/layout.py

```diff
@@ -0,0 +1,259 @@
+        self.dpath_scratch: Path
+        self.dpath_raw_dicom: Path
+        self.dpath_logs: Path
+        self.dpath_tabular: Path
+        self.dpath_assessments: Path
+        self.dpath_demographics: Path
+
+        # files (for type hinting)
+        self.fpath_config: Path
+        self.fpath_doughnut: Path
+        self.fpath_manifest: Path
+        self.fpath_imaging_bagel: Path
+
+        # directory names
+        self.dname_pipeline_work = "work"
+        self.dname_pipeline_output = "output"
+
+    def get_full_path(self, path: str | Path) -> Path:
+        """Build a full path from a relative path."""
+        return self.dpath_root / path
+
+    def __getattribute__(self, name: str) -> Any:
+        try:
+            return super().__getattribute__(name)
+        except AttributeError as exception:
+            if name in self.config.path_labels:
+                return self.get_full_path(self.config.get_path_info(name).path)
+            else:
+                raise exception
+
+    def get_paths(self, directory=True, include_optional=False) -> list[Path]:
+        """Return a list of all directory or file paths."""
+        paths = [
+            self.get_full_path(path_info.path)
+            for path_info in self.config.path_infos
+            if directory == path_info._is_directory
+            and (include_optional or path_info._is_required)
+        ]
+        return paths
+
+    @cached_property
+    def dpaths(self) -> list[Path]:
+        """Return a list of all required directory paths."""
+        return self.get_paths(directory=True)
+
+    @cached_property
+    def fpaths(self) -> list[Path]:
+        """Return a list of all required file paths."""
+        return self.get_paths(directory=False)
+
+    @cached_property
+    def dpath_descriptions(self) -> list[Tuple[Path, str]]:
+        """Return a list of directory paths and associated description strings."""
+        info_list = [
+            (self.get_full_path(path_info.path), path_info.description)
+            for path_info in self.config.path_infos
+            if path_info._is_directory and path_info.description is not None
+        ]
+        return info_list
+
+    def _find_missing_paths(self) -> list[Path]:
+        """Return a list of missing paths."""
+        missing = [dpath for dpath in self.dpaths if not dpath.exists()]
+        for fpath in self.fpaths:
+            if not fpath.exists():
+                missing.append(fpath)
+        return missing
+
+    def validate(self) -> bool:
+        """Validate that all the expected paths exist."""
+        missing_paths = self._find_missing_paths()
+        if len(missing_paths) != 0:
+            raise FileNotFoundError(
+                f"Missing {len(missing_paths)} paths"
+                f": {[str(path) for path in missing_paths]}"
+            )
+        return True
+
+    def get_dpath_pipeline(self, pipeline_name: str, pipeline_version: str) -> Path:
```

Comment:
So this part of BIDS is recommended, which in BIDS validator language that maybe one day the validator may throw  a warning about this.

However I don't think that the validator will start doing nested validation of BIDS dataset anytime soon.

In this case it is not even nested validation because the derivatives are not in the BIDS rawdata folder, so this is even less of a worry.

The only advantage I see at using this `<pipeline>-<variant>` convention instead of what nipoppy has been doing so far `<pipeline>/<variant>` is what was mentioned in a meeting this week: runnning the same pipeline + variant but with different parameters. This would probably involve adding another level of subfolder.

So it will depend on how frequently you expect to have people running different variants or parameters set.

Practically that would also mean renaming / moving folders in datasets that you have already processed. Not huge, but maybe annoying.

So my short answer would be: keep things like you had it until now, but open an issue to think if you make the output layout configurable by allowing something like:

```json
 "output_layout": "<pipeline>/<variant>",
 "output_layout": "<pipeline>-<variant>",
 "output_layout": "<pipeline>-<variant>/<parameters>"
```